File: chronik/chronik-plugin-impl/tests/test_tx_to_py.rs
// Copyright (c) 2024 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

use std::collections::BTreeMap;

use abc_rust_error::Result;
use bitcoinsuite_core::{
    hash::ShaRmd160,
    script::Script,
    tx::{Coin, OutPoint, Tx, TxId, TxInput, TxMut, TxOutput},
};
use bitcoinsuite_slp::{
    alp::{
        burn_section, genesis_section, mint_section, sections_opreturn,
        send_section,
    },
    color::ColoredTx,
    parsed::ParsedMintData,
    slp::{burn_opreturn, genesis_opreturn, mint_opreturn, send_opreturn},
    structs::{GenesisInfo, Token, TokenVariant},
    test_helpers::{
        meta_alp, meta_alp_unknown, meta_slp, spent_amount, spent_amount_group,
        spent_baton, EMPTY_TOKEN_ID, TOKEN_ID2, TOKEN_ID3, TOKEN_ID4,
        TOKEN_ID5, TOKEN_ID6, TOKEN_ID7, TOKEN_ID8,
    },
    token_type::{AlpTokenType, SlpTokenType},
    verify::{SpentToken, VerifyContext},
};
use chronik_plugin_common::data::{
    PluginNameMap, PluginOutput, PluginOutputEntry,
};
use chronik_plugin_impl::{
    module::{add_test_framework_to_pythonpath, load_plugin_module},
    tx::TxModule,
};
use pyo3::{
    types::{PyAnyMethods, PyModule},
    PyErr, PyObject, PyResult, Python,
};

#[derive(Clone, Debug, Default)]
struct MakePyTxParams<'a> {
    txid_num: u8,
    num_outputs: usize,
    op_return_script: Script,
    spent_tokens: &'a [Option<SpentToken>],
    plugin_outputs: &'a [Option<PluginOutput>],
    plugin_names: &'a [&'a str],
}

fn make_py_tx(
    py: Python<'_>,
    tx_module: &TxModule,
    params: MakePyTxParams<'_>,
) -> PyResult<PyObject> {
    let tx = Tx::with_txid(
        TxId::from([params.txid_num; 32]),
        TxMut {
            version: 0,
            inputs: params
                .spent_tokens
                .iter()
                .enumerate()
                .map(|(idx, _)| TxInput {
                    prev_out: OutPoint {
                        txid: TxId::from([2; 32]),
                        out_idx: idx as u32,
                    },
                    coin: Some(Coin::default()),
                    ..Default::default()
                })
                .collect(),
            outputs: [TxOutput {
                value: 0,
                script: params.op_return_script.clone(),
            }]
            .into_iter()
            .chain((0..params.num_outputs).map(|_| TxOutput::default()))
            .collect(),
            locktime: 0,
        },
    );
    let plugin_outputs = params
        .plugin_outputs
        .iter()
        .zip(&tx.inputs)
        .filter_map(|(plugin_output, input)| {
            let plugin_output = plugin_output.as_ref()?;
            Some((input.prev_out, plugin_output.clone()))
        })
        .collect::<BTreeMap<_, _>>();
    let colored_tx = ColoredTx::color_tx(&tx).unwrap_or_else(|| ColoredTx {
        outputs: vec![None; tx.outputs.len()],
        ..Default::default()
    });
    let context = VerifyContext {
        spent_tokens: params.spent_tokens,
        spent_scripts: None,
        genesis_info: None,
        override_has_mint_vault: None,
    };
    let token_tx = context.verify(colored_tx);
    tx_module.bridge_tx(
        py,
        &tx,
        Some((&token_tx, params.spent_tokens)),
        &plugin_outputs,
        &PluginNameMap::new(params.plugin_names.iter().enumerate().map(
            |(plugin_idx, plugin_name)| {
                (plugin_idx as u32, plugin_name.to_string())
            },
        )),
    )
}

#[test]
fn test_tx_to_py() -> Result<()> {
    let non_token_tx = Tx::with_txid(
        TxId::new([1; 32]),
        TxMut {
            version: 1,
            inputs: vec![
                TxInput {
                    prev_out: OutPoint {
                        txid: TxId::new([5; 32]),
                        out_idx: 7,
                    },
                    script: Script::new(b"\x01\x01".as_ref().into()),
                    sequence: 0x12345678,
                    coin: Some(Coin {
                        output: TxOutput {
                            script: Script::p2sh(&ShaRmd160([2; 20])),
                            value: 50000,
                        },
                        height: 0,
                        is_coinbase: false,
                    }),
                },
                TxInput {
                    prev_out: OutPoint {
                        txid: TxId::new([8; 32]),
                        out_idx: 22,
                    },
                    ..Default::default()
                },
            ],
            outputs: vec![TxOutput {
                value: 40000,
                script: Script::p2pkh(&ShaRmd160([6; 20])),
            }],
            locktime: 0x87654321,
        },
    );

    load_plugin_module();
    pyo3::prepare_freethreaded_python();
    Python::with_gil(|py| -> Result<_> {
        add_test_framework_to_pythonpath(py)?;

        let handle_exc = |py_err: PyErr| {
            py_err.print(py);
            py_err
        };

        let tx_module = TxModule::import(py)?;

        let test_module = PyModule::from_code_bound(
            py,
            include_str!("test_tx_to_py.py"),
            "test_tx_to_py.py",
            "test_tx_to_py",
        )?;

        test_module
            .getattr("test_non_token_tx")?
            .call1((tx_module.bridge_tx(
                py,
                &non_token_tx,
                None,
                &BTreeMap::new(),
                &Default::default(),
            )?,))
            .map_err(handle_exc)?;

        let slp_genesis_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 2,
                num_outputs: 2,
                op_return_script: genesis_opreturn(
                    &GenesisInfo {
                        token_ticker: b"SLP FUNGIBLE".as_ref().into(),
                        token_name: b"Slp Fungible".as_ref().into(),
                        mint_vault_scripthash: None,
                        url: b"https://slp.fungible".as_ref().into(),
                        hash: Some([b'x'; 32]),
                        data: None,
                        auth_pubkey: None,
                        decimals: 4,
                    },
                    SlpTokenType::Fungible,
                    Some(2),
                    1234,
                ),
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_slp_genesis_tx")?
            .call1((slp_genesis_tx,))
            .map_err(handle_exc)?;

        let slp_mint_vault_genesis_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 2,
                num_outputs: 1,
                op_return_script: genesis_opreturn(
                    &GenesisInfo {
                        token_ticker: b"SLP MINT VAULT".as_ref().into(),
                        token_name: b"Slp Mint Vault".as_ref().into(),
                        mint_vault_scripthash: Some(ShaRmd160([5; 20])),
                        url: b"https://slp.mintvault".as_ref().into(),
                        hash: None,
                        data: None,
                        auth_pubkey: None,
                        decimals: 4,
                    },
                    SlpTokenType::MintVault,
                    None,
                    1234,
                ),
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_slp_mint_vault_genesis_tx")?
            .call1((slp_mint_vault_genesis_tx,))
            .map_err(handle_exc)?;

        let slp_nft1_child_genesis_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 2,
                num_outputs: 1,
                op_return_script: genesis_opreturn(
                    &GenesisInfo::empty_slp(),
                    SlpTokenType::Nft1Child,
                    None,
                    1,
                ),
                spent_tokens: &[spent_amount(
                    meta_slp(TOKEN_ID3, SlpTokenType::Nft1Group),
                    1,
                )],
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_slp_nft1_child_genesis_tx")?
            .call1((slp_nft1_child_genesis_tx,))
            .map_err(handle_exc)?;

        let slp_mint_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 2,
                num_outputs: 2,
                op_return_script: mint_opreturn(
                    &TOKEN_ID3,
                    SlpTokenType::Fungible,
                    Some(2),
                    1234,
                ),
                spent_tokens: &[spent_baton(meta_slp(
                    TOKEN_ID3,
                    SlpTokenType::Fungible,
                ))],
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_slp_mint_tx")?
            .call1((slp_mint_tx,))
            .map_err(handle_exc)?;

        let slp_send_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 2,
                num_outputs: 3,
                op_return_script: send_opreturn(
                    &TOKEN_ID3,
                    SlpTokenType::Fungible,
                    &[5, 6, 7],
                ),
                spent_tokens: &[spent_amount(
                    meta_slp(TOKEN_ID3, SlpTokenType::Fungible),
                    20,
                )],
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_slp_send_tx")?
            .call1((slp_send_tx,))
            .map_err(handle_exc)?;

        let slp_burn_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 2,
                num_outputs: 0,
                op_return_script: burn_opreturn(
                    &TOKEN_ID3,
                    SlpTokenType::Fungible,
                    500,
                ),
                spent_tokens: &[spent_amount(
                    meta_slp(TOKEN_ID3, SlpTokenType::Fungible),
                    600,
                )],
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_slp_burn_tx")?
            .call1((slp_burn_tx,))
            .map_err(handle_exc)?;

        let alp_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 1,
                num_outputs: 10,
                op_return_script: sections_opreturn(vec![
                    genesis_section(
                        AlpTokenType::Standard,
                        &GenesisInfo {
                            token_ticker: b"ALP STANDARD".as_ref().into(),
                            token_name: b"Alp Standard".as_ref().into(),
                            mint_vault_scripthash: None,
                            url: b"https://alp.std".as_ref().into(),
                            hash: None,
                            data: Some(b"ALP DATA".as_ref().into()),
                            auth_pubkey: Some(b"ALP PubKey".as_ref().into()),
                            decimals: 2,
                        },
                        &ParsedMintData {
                            amounts: vec![0, 0, 10, 0, 0],
                            num_batons: 2,
                        },
                    ),
                    mint_section(
                        &TOKEN_ID2,
                        AlpTokenType::Standard,
                        &ParsedMintData {
                            amounts: vec![1000, 0, 0],
                            num_batons: 1,
                        },
                    ),
                    send_section(
                        &TOKEN_ID3,
                        AlpTokenType::Standard,
                        vec![0, 500, 0, 0, 0, 0, 0, 0, 6000],
                    ),
                    burn_section(&TOKEN_ID3, AlpTokenType::Standard, 1000),
                    send_section(
                        &TOKEN_ID8,
                        AlpTokenType::Standard,
                        vec![0, 0, 0, 0, 40],
                    ),
                    b"SLP2\x02".as_ref().into(),
                ]),
                spent_tokens: &[
                    spent_baton(meta_alp(TOKEN_ID2)),
                    None,
                    spent_amount(meta_alp(TOKEN_ID3), 2000),
                    spent_amount(meta_alp(TOKEN_ID3), 5000),
                    spent_amount(
                        meta_slp(TOKEN_ID4, SlpTokenType::Fungible),
                        30,
                    ),
                    spent_amount(
                        meta_slp(TOKEN_ID5, SlpTokenType::MintVault),
                        20,
                    ),
                    spent_amount(
                        meta_slp(TOKEN_ID6, SlpTokenType::Nft1Group),
                        20,
                    ),
                    spent_amount_group(
                        meta_slp(TOKEN_ID7, SlpTokenType::Nft1Child),
                        1,
                        meta_slp(TOKEN_ID6, SlpTokenType::Nft1Group),
                    ),
                    Some(SpentToken {
                        token: Token {
                            meta: meta_alp_unknown(EMPTY_TOKEN_ID, 3),
                            variant: TokenVariant::Unknown(3),
                        },
                        group_token_meta: None,
                    }),
                    Some(SpentToken {
                        token: Token {
                            meta: meta_slp(
                                EMPTY_TOKEN_ID,
                                SlpTokenType::Unknown(3),
                            ),
                            variant: TokenVariant::Unknown(3),
                        },
                        group_token_meta: None,
                    }),
                ],
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_alp_tx")?
            .call1((alp_tx,))
            .map_err(handle_exc)?;

        let non_token_burn_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 1,
                num_outputs: 1,
                op_return_script: Script::default(),
                spent_tokens: &[spent_amount(meta_alp(TOKEN_ID2), 200)],
                ..Default::default()
            },
        )?;
        test_module
            .getattr("test_non_token_burn_tx")?
            .call1((non_token_burn_tx,))
            .map_err(handle_exc)?;

        let plugin_inputs_tx = make_py_tx(
            py,
            &tx_module,
            MakePyTxParams {
                txid_num: 1,
                num_outputs: 1,
                op_return_script: Script::default(),
                spent_tokens: &[None, None],
                plugin_outputs: &[
                    Some(PluginOutput {
                        plugins: vec![
                            (
                                0,
                                PluginOutputEntry {
                                    groups: vec![
                                        b"grp1".to_vec(),
                                        b"grp2".to_vec(),
                                    ],
                                    data: vec![b"dat1".to_vec()],
                                },
                            ),
                            (
                                1,
                                PluginOutputEntry {
                                    groups: vec![b"2grp".to_vec()],
                                    data: vec![b"2dat".to_vec()],
                                },
                            ),
                            (
                                // Entries without loaded plugin are skipped
                                1337,
                                PluginOutputEntry {
                                    groups: vec![b"doesntexist".to_vec()],
                                    data: vec![b"doesntexist".to_vec()],
                                },
                            ),
                        ]
                        .into_iter()
                        .collect(),
                    }),
                    None,
                ],
                plugin_names: &["plg1", "plg2"],
            },
        )?;
        test_module
            .getattr("test_plugin_inputs_tx")?
            .call1((plugin_inputs_tx,))
            .map_err(handle_exc)?;

        Ok(())
    })
}

File: chronik/chronik-plugin-impl/tests/test_tx_to_py.py
# Copyright (c) 2024 The Bitcoin developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.

from chronik_plugin.etoken import GenesisInfo, Token, TokenTxEntry
from chronik_plugin.script import CScript
from chronik_plugin.tx import OutPoint, PluginOutputEntry, Tx, TxInput, TxOutput
from test_framework.util import assert_equal


def slp_amount(token_id: str, token_type: int, amount: int, entry_idx=0) -> Token:
    return Token(
        token_id=token_id,
        token_protocol="SLP",
        token_type=token_type,
        entry_idx=entry_idx,
        amount=amount,
        is_mint_baton=False,
    )


def slp_baton(token_id: str, token_type: int, entry_idx=0) -> Token:
    return Token(
        token_id=token_id,
        token_protocol="SLP",
        token_type=token_type,
        entry_idx=entry_idx,
        amount=0,
        is_mint_baton=True,
    )


def alp_amount(token_id: str, token_type: int, amount: int, entry_idx=0) -> Token:
    return Token(
        token_id=token_id,
        token_protocol="ALP",
        token_type=token_type,
        entry_idx=entry_idx,
        amount=amount,
        is_mint_baton=False,
    )


def alp_baton(token_id: str, token_type: int, entry_idx=0) -> Token:
    return Token(
        token_id=token_id,
        token_protocol="ALP",
        token_type=token_type,
        entry_idx=entry_idx,
        amount=0,
        is_mint_baton=True,
    )


def test_non_token_tx(tx: Tx):
    assert_equal(tx.txid, b"\x01" * 32)
    assert_equal(tx.version, 1)
    assert_equal(
        tx.inputs,
        [
            TxInput(
                prev_out=OutPoint(b"\x05" * 32, 7),
                script=CScript(bytes.fromhex("0101")),
                output=TxOutput(
                    script=CScript(
                        bytes.fromhex("a914020202020202020202020202020202020202020287")
                    ),
                    value=50000,
                    token=None,
                ),
                sequence=0x12345678,
                plugin={},
            ),
            TxInput(
                prev_out=OutPoint(b"\x08" * 32, 22),
                script=CScript(b""),
                output=None,
                sequence=0,
                plugin={},
            ),
        ],
    )
    assert_equal(
        tx.outputs,
        [
            TxOutput(
                script=CScript(
                    bytes.fromhex("76a914060606060606060606060606060606060606060688ac")
                ),
                value=40000,
                token=None,
            ),
        ],
    )
    assert_equal(tx.lock_time, 0x87654321)
    assert_equal(tx.token_entries, [])
    assert_equal(tx.empp_data, [])


def test_slp_genesis_tx(tx: Tx):
    assert_equal(tx.txid, b"\x02" * 32)
    assert_equal(tx.inputs, [])
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
            slp_amount("02" * 32, 1, 1234),
            slp_baton("02" * 32, 1),
        ],
    )
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="02" * 32,
                token_protocol="SLP",
                token_type=1,
                tx_type="GENESIS",
                group_token_id=None,
                genesis_info=GenesisInfo(
                    token_ticker=b"SLP FUNGIBLE",
                    token_name=b"Slp Fungible",
                    mint_vault_scripthash=None,
                    url=b"https://slp.fungible",
                    hash=b"x" * 32,
                    data=None,
                    auth_pubkey=None,
                    decimals=4,
                ),
            ),
        ],
    )


def test_slp_mint_vault_genesis_tx(tx: Tx):
    assert_equal(tx.txid, b"\x02" * 32)
    assert_equal(tx.inputs, [])
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
            slp_amount("02" * 32, 2, 1234),
        ],
    )
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="02" * 32,
                token_protocol="SLP",
                token_type=2,
                tx_type="GENESIS",
                group_token_id=None,
                genesis_info=GenesisInfo(
                    token_ticker=b"SLP MINT VAULT",
                    token_name=b"Slp Mint Vault",
                    mint_vault_scripthash=b"\x05" * 20,
                    url=b"https://slp.mintvault",
                    hash=None,
                    data=None,
                    auth_pubkey=None,
                    decimals=4,
                ),
            ),
        ],
    )


def test_slp_nft1_child_genesis_tx(tx: Tx):
    assert_equal(tx.txid, b"\x02" * 32)
    assert_equal(
        [inpt.output.token for inpt in tx.inputs],
        [
            slp_amount("03" * 32, 0x81, 1, entry_idx=1),
        ],
    )
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
            slp_amount("02" * 32, 0x41, 1),
        ],
    )
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="02" * 32,
                token_protocol="SLP",
                token_type=0x41,
                tx_type="GENESIS",
                group_token_id="03" * 32,
                genesis_info=GenesisInfo(
                    token_ticker=b"",
                    token_name=b"",
                    mint_vault_scripthash=None,
                    url=b"",
                    hash=None,
                    data=None,
                    auth_pubkey=None,
                    decimals=0,
                ),
            ),
            TokenTxEntry(
                token_id="03" * 32,
                token_protocol="SLP",
                token_type=0x81,
            ),
        ],
    )


def test_slp_mint_tx(tx: Tx):
    assert_equal(tx.txid, b"\x02" * 32)
    assert_equal(
        [inpt.output.token for inpt in tx.inputs],
        [
            slp_baton("03" * 32, 1),
        ],
    )
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
            slp_amount("03" * 32, 1, 1234),
            slp_baton("03" * 32, 1),
        ],
    )
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="03" * 32,
                token_protocol="SLP",
                token_type=1,
                tx_type="MINT",
            ),
        ],
    )


def test_slp_send_tx(tx: Tx):
    assert_equal(tx.txid, b"\x02" * 32)
    assert_equal(
        [inpt.output.token for inpt in tx.inputs],
        [
            slp_amount("03" * 32, 1, 20),
        ],
    )
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
            slp_amount("03" * 32, 1, 5),
            slp_amount("03" * 32, 1, 6),
            slp_amount("03" * 32, 1, 7),
        ],
    )
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="03" * 32,
                token_protocol="SLP",
                token_type=1,
                tx_type="SEND",
                actual_burn_amount=2,
            ),
        ],
    )


def test_slp_burn_tx(tx: Tx):
    assert_equal(tx.txid, b"\x02" * 32)
    assert_equal(
        [inpt.output.token for inpt in tx.inputs],
        [
            slp_amount("03" * 32, 1, 600),
        ],
    )
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
        ],
    )
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="03" * 32,
                token_protocol="SLP",
                token_type=1,
                tx_type="BURN",
                actual_burn_amount=600,
                intentional_burn_amount=500,
            ),
        ],
    )


def test_alp_tx(tx: Tx):
    assert_equal(tx.txid, b"\x01" * 32)
    assert_equal(len(tx.token_entries), 11)
    assert_equal(
        tx.token_entries[0],
        TokenTxEntry(
            token_id="01" * 32,
            token_protocol="ALP",
            token_type=0,
            tx_type="GENESIS",
            genesis_info=GenesisInfo(
                token_ticker=b"ALP STANDARD",
                token_name=b"Alp Standard",
                mint_vault_scripthash=None,
                url=b"https://alp.std",
                hash=None,
                data=b"ALP DATA",
                auth_pubkey=b"ALP PubKey",
                decimals=2,
            ),
        ),
    )
    assert_equal(
        tx.token_entries[1],
        TokenTxEntry(
            token_id="02" * 32,
            token_protocol="ALP",
            token_type=0,
            tx_type="MINT",
        ),
    )
    assert_equal(
        tx.token_entries[2],
        TokenTxEntry(
            token_id="03" * 32,
            token_protocol="ALP",
            token_type=0,
            tx_type="SEND",
            actual_burn_amount=500,
            intentional_burn_amount=1000,
        ),
    )
    assert_equal(
        tx.token_entries[3],
        TokenTxEntry(
            token_id="08" * 32,
            token_protocol="ALP",
            token_type=0,
            tx_type="SEND",
            is_invalid=True,
        ),
    )
    assert_equal(
        tx.token_entries[4],
        TokenTxEntry(
            token_id="00" * 32,
            token_protocol="ALP",
            token_type=2,
            tx_type="UNKNOWN",
        ),
    )
    assert_equal(
        tx.token_entries[5],
        TokenTxEntry(
            token_id="00" * 32,
            token_protocol="SLP",
            token_type=3,
            is_invalid=True,
        ),
    )
    assert_equal(
        tx.token_entries[6],
        TokenTxEntry(
            token_id="00" * 32,
            token_protocol="ALP",
            token_type=3,
            is_invalid=True,
        ),
    )
    assert_equal(
        tx.token_entries[7],
        TokenTxEntry(
            token_id="04" * 32,
            token_protocol="SLP",
            token_type=1,
            is_invalid=True,
            actual_burn_amount=30,
        ),
    )
    assert_equal(
        tx.token_entries[8],
        TokenTxEntry(
            token_id="05" * 32,
            token_protocol="SLP",
            token_type=2,
            is_invalid=True,
            actual_burn_amount=20,
        ),
    )
    assert_equal(
        tx.token_entries[9],
        TokenTxEntry(
            token_id="06" * 32,
            token_protocol="SLP",
            token_type=0x81,
            is_invalid=True,
            actual_burn_amount=20,
        ),
    )
    assert_equal(
        tx.token_entries[10],
        TokenTxEntry(
            token_id="07" * 32,
            token_protocol="SLP",
            token_type=0x41,
            group_token_id="0606060606060606060606060606060606060606060606060606060606060606",
            is_invalid=True,
            actual_burn_amount=1,
        ),
    )
    assert_equal(
        [inpt.output.token for inpt in tx.inputs],
        [
            alp_baton("02" * 32, 0, entry_idx=1),
            None,
            alp_amount("03" * 32, 0, 2000, entry_idx=2),
            alp_amount("03" * 32, 0, 5000, entry_idx=2),
            slp_amount("04" * 32, 1, 30, entry_idx=7),
            slp_amount("05" * 32, 2, 20, entry_idx=8),
            slp_amount("06" * 32, 0x81, 20, entry_idx=9),
            slp_amount("07" * 32, 0x41, 1, entry_idx=10),
            alp_amount("00" * 32, 3, 0, entry_idx=6),
            slp_amount("00" * 32, 3, 0, entry_idx=5),
        ],
    )
    assert_equal(
        [output.token for output in tx.outputs],
        [
            None,
            alp_amount("02" * 32, 0, 1000, entry_idx=1),
            alp_amount("03" * 32, 0, 500, entry_idx=2),
            alp_amount("01" * 32, 0, 10, entry_idx=0),
            alp_baton("02" * 32, 0, entry_idx=1),
            None,
            alp_baton("01" * 32, 0, entry_idx=0),
            alp_baton("01" * 32, 0, entry_idx=0),
            alp_amount("00" * 32, 2, 0, entry_idx=4),
            alp_amount("03" * 32, 0, 6000, entry_idx=2),
            alp_amount("00" * 32, 2, 0, entry_idx=4),
        ],
    )


def test_non_token_burn_tx(tx: Tx):
    assert_equal(
        tx.token_entries,
        [
            TokenTxEntry(
                token_id="02" * 32,
                token_protocol="ALP",
                token_type=0,
                is_invalid=True,
                actual_burn_amount=200,
            )
        ],
    )
    assert_equal(
        [inpt.output.token for inpt in tx.inputs],
        [alp_amount("02" * 32, 0, 200)],
    )
    assert_equal(
        [output.token for output in tx.outputs],
        [None, None],
    )


def test_plugin_inputs_tx(tx: Tx):
    assert_equal(
        [inpt.plugin for inpt in tx.inputs],
        [
            {
                "plg1": PluginOutputEntry(groups=[b"grp1", b"grp2"], data=[b"dat1"]),
                "plg2": PluginOutputEntry(groups=[b"2grp"], data=[b"2dat"]),
            },
            {},
        ],
    )
